### PR TITLE
[rust-axum] Split up api trait per tag

### LINKED ISSUE #18442: [REQ] [Rust-Axum] Solution to split implementation across files
### Is your feature request related to a problem? Please describe.

The way the generator currently works means you are forced to implement all functions for the api in a single block. While this works for small simple projects, it becomes quite unwieldy for large apis.

## Describe the solution you'd like

I wonder if potentially the `Api` trait that is generated could instead be split up, perhaps for each endpoint. Each trait could contain just a single function. This would allow you to impl each endpoint seperately and thus allow them to be organised into seperate files. For example if you had an api that contained `foo`, `bar` and `baz`, rather than generating a single `Api` trait with all functions, it could instead generate `ApiFoo`, `ApiBar` and `ApiBaz`. `server::new()` could be trait bound to all of the generated traits rather than just the single `Api` trait which would still ensure the final server implements all the required functionality.

## Describe alternatives you've considered

Another option would be to potentially split the traits up by tags.

## Additional context

I have not attempted trait bounds at such a large scale so I am not sure if there are any limitations or overheads at such a scale. I feel like perhaps this would be best as a config option rather than the default.

### PR DESCRIPTION
**Pull request by kalvinpearce**
The rust-axum generator creates a single `Api` trait that holds all operations. This PR instead creates a trait for each tag. This allows for more flexibility in how the api is implemented as one can use the openapi normalizer to modify the tags.

cc @linxGnu

Fixes #18442.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by wing328** [comment]
please run ` ./bin/generate-samples.sh ./bin/configs/manual/rust-axum*` to update the samples (use git bash or wsl if you're using windows)

**Comment by kalvinpearce** [comment]
Thanks for that, I had forgotten to run the rust specific samples and also caught an issue with one of the use statements, fixed in [7d6dbfc](https://github.com/OpenAPITools/openapi-generator/pull/18621/commits/7d6dbfc346704eeb9238d7d77f31372d1d5c5975)

**Comment by kalvinpearce** [comment]
@linxGnu I have implemented [your proposal](https://github.com/OpenAPITools/openapi-generator/issues/18442#issuecomment-2103782279) of sorting and grouping traits to handle operations with multiple tags (see commit [a896ed8](https://github.com/OpenAPITools/openapi-generator/pull/18621/commits/a896ed821d016452971e15d501f1c5cb8227c859))

I'm not sure where the best place to implement it would be, however I have got it working currently in `addOperationToGroup`. Please let me know if there is a more fitting place to implement it and I would be happy to move it there instead.

**Comment by linxGnu** [comment]
Hi @kalvinpearce 

Thank you so much for great PR. I need sometime to review it. Hopefully will feedback to you around 11:00am Tue next week (Japan time)

**Comment by kalvinpearce** [comment]
No worries, please take your time. I am not a Java developer, nor have I looked too deeply into the inner workings of this codebase so please be as particular as you would like regarding code styling etc, I am happy to make any adjustments you recommend.

**Comment by wing328** [comment]
>  This allows for more flexibility in how the api is implemented as one can use the openapi normalizer to modify the tags.

Totally agreed with this one. We can add more normalizer rules to cater more use cases if needed.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustAxumServerCodegen.java [hunk 1]
```diff
@@ -649,13 +657,23 @@ public boolean isDataTypeFile(final String dataType) {
...
+            // Skip any tags other than the first one. This is because we
+            // override the tag with a combined version of all the tags.
             String expectedTag = sanitizeTag(op.tags.get(0).getName());
             if (!tag.equals(expectedTag)) {
                 LOGGER.info("generated skip additional tag `{}` with operationId={}", tag, op.operationId);
                 return;
             }
+
+            // Get all tags sorted by name
+            List<String> tags = op.tags.stream().map(t -> t.getName()).sorted().collect(Collectors.toList());
```
Comment (linxGnu) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustAxumServerCodegen.java:671]: Please add `final`:
```java
final  List<String> tags
```

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustAxumServerCodegen.java [hunk 2]
```diff
@@ -649,13 +657,23 @@ public boolean isDataTypeFile(final String dataType) {
...
+            // Combine into a single group
+            String combinedTag = tags.stream().collect(Collectors.joining("-"));
```
Comment (linxGnu) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustAxumServerCodegen.java:673]: Ditto

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustAxumServerCodegen.java [hunk 3]
```diff
@@ -565,6 +571,8 @@ public CodegenOperation fromOperation(String path, String httpMethod, Operation
     @Override
     public OperationsMap postProcessOperationsWithModels(OperationsMap operationsMap, List<ModelMap> allModels) {
         OperationMap operations = operationsMap.getOperations();
+        String classname = operations.getClassname();
+        operations.put("classnamePascalCase", camelize(classname));
```
Comment (linxGnu) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustAxumServerCodegen.java:575]: Please combine these lines into one:
```java
operations.put("classnamePascalCase", camelize(operations.getClassname()));
```

**Review by linxGnu — APPROVED** [review]
The output looks awesome to me! I just have some small comments, please fix them then LGTM @kalvinpearce

**Comment by kalvinpearce** [comment]
Thank you both for the feedback. @linxGnu I have implemented your comments in [01af1a8](https://github.com/OpenAPITools/openapi-generator/pull/18621/commits/01af1a8d07cbd1771b54c74165e7066f94faf0fd).

**Comment by linxGnu** [comment]
LGTM. Kindly cc @wing328 for help to review.

**Pull request merged**